src/bif/erlang.rs

```diff
@@ -0,0 +1,519 @@
+       _ => return Err(Exception::new(Reason::EXC_BADARG))
+    };
+    let heap = &process.context_mut().heap;
+    let tuple = value::tuple(&heap, num as u32);
+    for i in 0..num {
+        unsafe {
+            std::ptr::write(&mut tuple[i as usize], args[1].clone());
+        }
+    }
+    if !args[2].is_list() {
+        return Err(Exception::new(Reason::EXC_BADARG));
+    }
+    let init: &value::Cons = match args[2].try_into() {
+        Ok(cons) => cons,
+        _ => return Err(Exception::new(Reason::EXC_BADARG))
+    };
+    for item in init.iter() {
+        let t: &Tuple = match item.try_into() {
+            Ok(tuple) => tuple, // FIXME do the len checking here
+            _ => return Err(Exception::new(Reason::EXC_BADARG))
+        };
+        if t.len != 2 {
+            return Err(Exception::new(Reason::EXC_BADARG))
+        }
+        let n = match t[0].into_number() {
+            Ok(value::Num::Integer(i)) if ! i < 1 && i - 1 < num => i - 1,
+            _ => return Err(Exception::new(Reason::EXC_BADARG))
+        };
+        unsafe {
+            std::ptr::write(&mut tuple[n as usize], t[1]);
+        }
+    }
+    Ok(Term::from(tuple))
+}
+
+pub fn bif_erlang_append_element_2(_vm: &vm::Machine, process: &RcProcess, args: &[Term]) -> BifResult {
+    if !args[0].is_tuple() {
+        return Err(Exception::new(Reason::EXC_BADARG));
+    }
+    let t: &Tuple = match args[0].try_into() {
+        Ok(tuple) => tuple,
+        _ => return Err(Exception::new(Reason::EXC_BADARG))
+    };
+    let heap = &process.context_mut().heap;
+    let mut new_tuple = value::tuple(&heap, (t.len() + 1) as u32);
+    unsafe {
+        new_tuple[..t.len()].copy_from_slice(&t[..]);
+        std::ptr::write(&mut new_tuple[t.len()], args[1]);
+    }
+    Ok(Term::from(new_tuple))
+}
+
+pub fn bif_erlang_setelement_3(_vm: &vm::Machine, process: &RcProcess, args: &[Term]) -> BifResult {
+    let number = match args[0].into_number() {
+        Ok(value::Num::Integer(i)) if ! i < 1 => i - 1,
+        _ => return Err(Exception::new(Reason::EXC_BADARG))
+    };
+    let t: &Tuple = match args[1].try_into() {
+        Ok(tuple) => tuple,
+        _ => return Err(Exception::new(Reason::EXC_BADARG))
+    };
+    if !(number <= t.len() as i32) {
+        return Err(Exception::new(Reason::EXC_BADARG));
+    }
+    let heap = &process.context_mut().heap;
+    let mut new_tuple = value::tuple(&heap, t.len() as u32);
+    unsafe {
+        new_tuple[..t.len()].copy_from_slice(&t[..]);
+        std::ptr::write(&mut new_tuple[number as usize], args[2]);
+    }
+    Ok(Term::from(new_tuple))
+}
+
+pub fn bif_erlang_tuple_to_list_1(_vm: &vm::Machine, process: &RcProcess, args: &[Term]) -> BifResult {
+    if !args[0].is_tuple() {
+        return Err(Exception::new(Reason::EXC_BADARG));
+    }
+    let t: &Tuple = match args[0].try_into() {
+        Ok(tuple) => tuple,
+        _ => return Err(Exception::new(Reason::EXC_BADARG))
+    };
+    let mut cons = vec![];
```

Comment:
Vec is wasteful since it's dynamically allocated off the heap.

edit: actually, compared to the ETF solution I linked to you, there's a much easier solution for this specific BIF, just build the list backwards:

```rust
let mut n = tuple.len() - 1;
let mut list = Term::nil();

while n >= 0 {
    list = cons!(heap, tuple[n], list)
    n -= 1;
}

Ok(list)
```

That's how BEAM does it too: https://github.com/erlang/otp/blob/260935595647eb2c4ca94c4c853c2b6065610272/erts/emulator/beam/bif.c#L3394-L3397

Reply:
Update: for your other uses of `list_into_iter!` in the tests, I now implemented `from_iter()` for Cons: https://github.com/archseer/enigma/commit/a3a1d4f9ee1b19582901f2b9649d9328552cce11

So now we can replace `list_into_iter!` with `from_iter`. I will probably make the function a bit better in the future (I'm missing some iter impls for Tuple, etc.), but should be usable enough to finish this PR